test/backend-cpp/TestCategoryMapper.cpp

```diff
@@ -123,59 +156,106 @@ class CategoryMapperTester {
     module.push_back(funcOp);
   }
 
-  static bool verifyFunction(OMTensor *out, OMTensor *expected) {
-    // Verify that the output tensor has the expected rank/extents.
-    if (omTensorGetRank(out) != omTensorGetRank(expected)) {
-      llvm::errs() << "Output tensor has rank " << omTensorGetRank(out)
-                   << ", expecting " << omTensorGetRank(expected) << "\n";
+  // Verify that the output tensor has the expected rank.
+  static bool verifyRank(const OMTensor &out, int64_t rank) {
+    if (omTensorGetRank(&out) != rank) {
+      llvm::errs() << "Output tensor has rank " << omTensorGetRank(&out)
+                   << ", expecting " << rank << "\n";
       return false;
     }
-    if (omTensorGetNumElems(out) != omTensorGetNumElems(expected)) {
-      llvm::errs() << "Output tensor has " << omTensorGetNumElems(out)
-                   << "elements, expecting " << omTensorGetNumElems(expected)
-                   << "\n";
+    return true;
+  }
+
+  // Verify that the output tensor has the expected number of elements.
+  static bool verifyNumElements(const OMTensor &out, int64_t numElems) {
+    if (omTensorGetNumElems(&out) != numElems) {
+      llvm::errs() << "Output tensor has " << omTensorGetNumElems(&out)
+                   << " elements, expecting " << numElems << "\n";
       return false;
     }
+    return true;
+  }
+
+  template <typename T>
+  static bool compareEqual(T val, T expectedVal) {
+    return val == expectedVal;
+  }
+
+  // Verification function.
+  // This function will be called back by the ModelBuilder.
+  template <typename T>
+  static bool verifyResults(const OMTensor *out, const OMTensor *expected) {
+    if (!verifyRank(*out, omTensorGetRank(expected)))
+      return false;
+    if (!verifyNumElements(*out, omTensorGetNumElems(expected)))
+      return false;
 
     // Verify that the output tensor contains the expected result.
-    auto outDataPtr = (const char **)omTensorGetDataPtr(out);
-    auto expectedDataPtr = (const char **)(omTensorGetDataPtr(expected));
+    const auto *outDataPtr = static_cast<T *>(omTensorGetDataPtr(out));
+    const auto *expDataPtr = static_cast<T *>(omTensorGetDataPtr(expected));
```

Comment:
Can you comment on if that cast is safer/ok compared to what was done in numerical, namely
```
  for (int64_t i = 0; i < I; ++i) {
    for (int64_t j = 0; j < J; ++j) {
      omTensorGetElem<float>(ref, {i, j}) = 0;
      for (int64_t k = 0; k < K; k++) {
        float aVal, bVal;
        if (aTrans == 0)
          aVal = omTensorGetElem<float>(a.get(), {i, k});
        else
          aVal = omTensorGetElem<float>(a.get(), {k, i});
        if (bTrans == 0)
          bVal = omTensorGetElem<float>(b.get(), {k, j});
        else
          bVal = omTensorGetElem<float>(b.get(), {j, k});
        omTensorGetElem<float>(ref, {i, j}) += aVal * bVal;
      }
    }
  }
```

I agree that the quoted code above is dreadfully slow by creating a vector for each accesses

Reply:
The implementation of `omTensorGetElem` is: 
```
template <typename T>
T &omTensorGetElem(const OMTensor *omt, std::vector<int64_t> indexes) {
  int64_t elemOffset = omTensorComputeElemOffset(omt, indexes);
  return ((T *)omt->_alignedPtr)[elemOffset];
}
```

The type cast on the return stmt is in C style. A C-style cast `(T *)` is **not** equivalent to a C++ `static_cast<T*>()` because a C-style cast is not type safe, while a `static_cast ` allows the compiler to report errors. For example: 

```
void foo(char *p) {
  int *pi = (int *)p;   // no compiler error here
  *pi = 3;              // ooops: just overwrote 3 extra bytes in mem. pointed to by p!!!
}
```

If a `static_cast<int*>(p)` was used the compiler would catch the potential runtime problem at compile time ==> safer!

If the intention was to "reinterpret" the memory pointed to by `p` then one should use a C++ `reintepret_cast`.

Of course all of this implies the code section is compiled by a C++ compiler.

Reply:
We prob. should modify the runtime implementation of `omTensorGetElem` and use C++ style casts. But that is another PR.

Reply:
ok